### ISSUE TITLE
Got http status code 500, when the node containing a manager master is unreachable

### ISSUE THREAD

**Issue by Argentum88** [leo-project/leofs#665]
I have two nodes
- node1: manager_0(master)/gatewey_0/storage_0
- node2: manager_1/gatewey_1/storage_1

step by step
1) put 10 small files (200kb)
2) disable network on node1
3) get key from gatewey_1  (code 200) all good
4) waiting 10-20 minutes
5) get key from gatewey_1  (code 500)
6) enable network on node1
7) all good

I also tried stop manager_0 instead disable network on node1, but still after a short time I got http status code 500

My environment:

* method communication with leofs: put operation - s3, get operation - http
* vagrant with virtualbox provider
* box ubuntu 16.04
* leofs was installed from the deb package (https://github.com/leo-project/leofs/releases/download/1.3.2.1/leofs_1.3.2.1-1_erl-18.3_ubuntu-16.04_amd64.deb)

```
ubuntu@node1:~$ leofs-adm status
 [System Confiuration]
-----------------------------------+----------
 Item                              | Value    
-----------------------------------+----------
 Basic/Consistency level
-----------------------------------+----------
                    system version | 1.3.2
                        cluster Id | leofs_1
                             DC Id | dc_1
                    Total replicas | 2
          number of successes of R | 1
          number of successes of W | 1
          number of successes of D | 1
 number of rack-awareness replicas | 0
                         ring size | 2^128
-----------------------------------+----------
 Multi DC replication settings
-----------------------------------+----------
        max number of joinable DCs | 2
           number of replicas a DC | 1
-----------------------------------+----------
 Manager RING hash
-----------------------------------+----------
                 current ring-hash | b69a2a25
                previous ring-hash | b69a2a25
-----------------------------------+----------

 [State of Node(s)]
-------+------------------------------+--------------+----------------+----------------+----------------------------
 type  |             node             |    state     |  current ring  |   prev ring    |          updated at         
-------+------------------------------+--------------+----------------+----------------+----------------------------
  S    | storage_0@192.168.33.10      | running      | b69a2a25       | b69a2a25       | 2017-04-01 03:44:35 +0000
  S    | storage_1@192.168.33.11      | running      | b69a2a25       | b69a2a25       | 2017-04-01 03:46:03 +0000
  G    | gateway_0@192.168.33.10      | running      | b69a2a25       | b69a2a25       | 2017-04-01 03:44:42 +0000
  G    | gateway_1@192.168.33.11      | running      | b69a2a25       | b69a2a25       | 2017-04-01 03:46:17 +0000
-------+------------------------------+--------------+----------------+----------------+----------------------------

```

gateway error log after http status code 500
```
[E]     gateway_1@192.168.33.11 2017-04-01 12:45:45.407556 +0000        1491050745      null:null       0       ** Node 'gateway_0@192.168.33.10' not responding **
** Removing (timedout) connection **
^M
[E]     gateway_1@192.168.33.11 2017-04-01 12:46:00.270673 +0000        1491050760      null:null       0       ** Node 'manager_0@192.168.33.10' not responding **
** Removing (timedout) connection **
^M
[E]     gateway_1@192.168.33.11 2017-04-01 12:46:00.271405 +0000        1491050760      null:null       0       ** Node 'storage_0@192.168.33.10' not responding **
** Removing (timedout) connection **

[E]     gateway_1@192.168.33.11 2017-04-01 13:08:43.259054 +0000        1491052123      null:null       0       Error in process <0.30116.3> on node 'gateway_1@192.168.33.11' with exit value:
{[{reason,{case_clause,{value,{badrpc,nodedown}}}},{mfa,{leo_gateway_s3_api,handle,2}},{stacktrace,[{leo_s3_bucket,find_bucket_by_name_2,3,[{file,"src/leo_s3_bucket.erl"},{line,948}]},{lists,foldl,3,[{file,"lists.erl"},{line,1262}]},{leo_s3_bucket,find_bucket_by_name_1,3,[{file,"src/leo_s3_bucket.erl"},{line,907}]},{leo_s3_bucket,get_latest_bucket,1,[{file,"src/leo_s3_bucket.erl"},{line,605}]},{leo_gateway_s3_api,handle,2,[{file,"src/leo_gateway_s3_api.erl"},{line,1425}]},{cowboy_handler,handler_handle,4,[{file,"src/cowboy_handler.erl"},{line,111}]},{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,442}]}]},{req,[{socket,#Port<0.6927>},{transport,ranch_tcp},{connection,keepalive},{pid,<0.30116.3>},{method,<<"GET">>},{version,'HTTP/1.1'},{peer,{{10,0,2,2},54298}},{host,<<"localhost">>},{host_info,undefined},{port,9081},{path,<<"/test_bucket/151159452958df25f92be731.60501008.png">>},{path_info,undefined},{qs,<<>>},{qs_vals,undefined},{bindings,[]},{headers,[{<<"host">>,<<"localhost:9081">>},{<<"connection">>,<<"keep-alive">>},{<<"cache-control">>,<<"max-age=0">>},{<<"upgrade-insecure-requests">>,<<"1">>},{<<"user-agent">>,<<"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.110 Safari/537.36">>},{<<"accept">>,<<"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8">>},{<<"accept-encoding">>,<<"gzip, deflate, sdch, br">>},{<<"accept-language">>,<<"ru-RU,ru;q=0.8,en-US;q=0.6,en;q=0.4">>},{<<"cookie">>,<<"_ym_uid=1488530582853988558; adminer_permanent=c2VydmVy-bXlzcWw%3D-cm9vdA%3D%3D-%3Afx3Lf6bskv4%3D">>},{<<"if-none-match">>,<<"\"bbe10cc8f64994c08c0ece3b8a4c8988\"">>},{<<"if-modified-since">>,<<"Sat, 01 Apr 2017 04:00:56 GMT">>}]},{p_headers,[{<<"connection">>,[<<"keep-alive">>]}]},{cookies,undefined},{meta,[]},{body_state,waiting},{buffer,<<>>},{multipart,undefined},{resp_compress,false},{resp_state,waiting},{resp_headers,[]},{resp_body,<<>>},{onresponse,undefined}]},{state,[{1,64},true,undefined,{http_options,leo_gateway_s3_api,8080,8443,"./etc/server_cert.pem","./etc/server_key.pem",128,4096,"./etc/http_custom_header.conf",5000,15000,5242880,inner,16,268435456,0,1048576,"./cache/data","./cache/journal",300,1048576,[],[],1000,5242880000,5242880,5242880,5767168}]}],[{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,442}]}]}
^M
[E]     gateway_1@192.168.33.11 2017-04-01 13:08:43.262409 +0000        1491052123      null:null       0       Ranch listener leo_gateway_s3_api had connection process started with cowboy_protocol:start_link/4 at <0.30116.3> exit with reason: {[{reason,{case_clause,{value,{badrpc,nodedown}}}},{mfa,{leo_gateway_s3_api,handle,2}},{stacktrace,[{leo_s3_bucket,find_bucket_by_name_2,3,[{file,"src/leo_s3_bucket.erl"},{line,948}]},{lists,foldl,3,[{file,"lists.erl"},{line,1262}]},{leo_s3_bucket,find_bucket_by_name_1,3,[{file,"src/leo_s3_bucket.erl"},{line,907}]},{leo_s3_bucket,get_latest_bucket,1,[{file,"src/leo_s3_bucket.erl"},{line,605}]},{leo_gateway_s3_api,handle,2,[{file,"src/leo_gateway_s3_api.erl"},{line,1425}]},{cowboy_handler,handler_handle,4,[{file,"src/cowboy_handler.erl"},{line,111}]},{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,442}]}]},{req,[{socket,#Port<0.6927>},{transport,ranch_tcp},{connection,keepalive},{pid,<0.30116.3>},{method,<<"GET">>},{version,'HTTP/1.1'},{peer,{{10,0,2,2},54298}},{host,<<"localhost">>},{host_info,undefined},{port,9081},{path,<<"/test_bucket/151159452958df25f92be731.60501008.png">>},{path_info,undefined},{qs,<<>>},{qs_vals,undefined},{bindings,[]},{headers,[{<<"host">>,<<"localhost:9081">>},{<<"connection">>,<<"keep-alive">>},{<<"cache-control">>,<<"max-age=0">>},{<<"upgrade-insecure-requests">>,<<"1">>},{<<"user-agent">>,<<"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.110 Safari/537.36">>},{<<"accept">>,<<"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8">>},{<<"accept-encoding">>,<<"gzip, deflate, sdch, br">>},{<<"accept-language">>,<<"ru-RU,ru;q=0.8,en-US;q=0.6,en;q=0.4">>},{<<"cookie">>,<<"_ym_uid=1488530582853988558; adminer_permanent=c2VydmVy-bXlzcWw%3D-cm9vdA%3D%3D-%3Afx3Lf6bskv4%3D">>},{<<"if-none-match">>,<<"\"bbe10cc8f64994c08c0ece3b8a4c8988\"">>},{<<"if-modified-since">>,<<"Sat, 01 Apr 2017 04:00:56 GMT">>}]},{p_headers,[{<<"connection">>,[<<"keep-alive">>]}]},{cookies,undefined},{meta,[]},{body_state,waiting},{buffer,<<>>},{multipart,undefined},{resp_compress,false},{resp_state,waiting},{resp_headers,[]},{resp_body,<<>>},{onresponse,undefined}]},{state,[{1,64},true,undefined,{http_options,leo_gateway_s3_api,8080,8443,"./etc/server_cert.pem","./etc/server_key.pem",128,4096,"./etc/http_custom_header.conf",5000,15000,5242880,inner,16,268435456,0,1048576,"./cache/data","./cache/journal",300,1048576,[],[],1000,5242880000,5242880,5242880,5767168}]}],[{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,442}]}]}
```

**Comment by Argentum88** [comment]
node1 configuration 
[node1.zip](https://github.com/leo-project/leofs/files/887712/node1.zip)

node2 configuration 
[node2.zip](https://github.com/leo-project/leofs/files/887713/node2.zip)

**Comment by yosukehara** [comment]
@Argentum88 This issue was already fixed with #642 , then we have been investigating LeoFS' performance during LeoManger' node down: #651 .

Thanks

**Comment by Argentum88** [comment]
I understand correctly that the patch will go into the next release?

**Comment by yosukehara** [comment]
@Argentum88 
>  I understand correctly that this fix will go into the next release?

Right, [v1.3.3](https://github.com/leo-project/leofs/milestone/65).

**Comment by windkit** [comment]
You can test it by upgrading the leo_s3_libs to the newest develop branch

```
cd deps/leo_s3_libs
git pull
git checkout develop
```

**Comment by yosukehara** [comment]
@Argentum88 Just in case, we will test [v1.3.3-dev](https://github.com/leo-project/leofs/tree/develop) on our environment whether this error does not appear with the current development version.

> 1. put 10 small files (200kb)
> 2. disable network on node1
> 3. get key from gatewey_1 (code 200) all good
> 4. waiting 10-20 minutes
> 5. get key from gatewey_1 (code 500)
> 6. enable network on node1
> 7. all good

**Comment by windkit** [comment]
I have confirmed this issue at 1.3.2.1

This issues stems from not handling `{badrpc, Reason}` in `leo_s3_bucket:find_bucket_by_name_2`
https://github.com/leo-project/leo_s3_libs/blob/1.2.9/src/leo_s3_bucket.erl#L939

which has been fixed after dialyzer raised a warning about that
https://github.com/leo-project/leo_s3_libs/blob/develop/src/leo_s3_bucket.erl#L971

So while this is not the same as #642, it will also be fixed by 1.3.3 release.

**Comment by yosukehara** [comment]
We confirmed this issue was fixed. You can test it after retrieving the latest develop branch and upgrading the leo_s3_libs to the newest develop branch:

```bash
$ cd deps/leo_s3_libs
$ git pull
git checkout develop
```
